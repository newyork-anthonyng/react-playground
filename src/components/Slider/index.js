import React, { Component } from "react";
import T from "prop-types";
import noop from "../../utility/noop";
import glamorous from "glamorous";

export const Rail = glamorous.div({
  backgroundColor: "#eee",
  border: "1px solid #888",
  height: "4px",
  position: "relative",
  width: "300px"
});

export const Knob = glamorous.div({
  backgroundColor: "#DDD",
  border: "1px solid #888",
  cursor: "pointer",
  height: "28px",
  position: "absolute",
  top: "-14px",
  width: "8px"
});

class Slider extends Component {
  constructor() {
    super();

    this.state = {
      offset: 0
    };

    this.isDragging = false;
    this.x = null;
  }

  componentDidMount() {
    window.addEventListener("mousemove", this.handleDragMove);
    window.addEventListener("mouseup", this.handleDragEnd);

    this.railWidth = this.rail.getBoundingClientRect().width;
  }

  componentWillUnmount() {
    window.removeEventListener("mousemove", this.handleDragMove);
    window.removeEventListener("mouseup", this.handleDragEnd);
  }

  handleDragStart = e => {
    this.isDragging = true;

    const event = (e.touches && e.touches[0]) || e;
    const clientX = event.clientX;

    this.x = event.clientX;
  };

  handleDragMove = e => {
    if (!this.isDragging) return;

    const event = (e.touches && e.touches[0]) || e;
    const clientX = event.clientX;

    const offset = Math.min(
      this.railWidth,
      Math.max(0, this.state.offset + (clientX - this.x))
    );
    this.setState({ offset });

    this.x = clientX;
  };

  handleDragEnd = () => {
    if (!this.isDragging) return;

    this.isDragging = false;
    this.x = null;
  };

  handleKeyDown = e => {
    // pressed UP or RIGHT arrows
    if (/(38|39)/.test(e.keyCode)) {
      this.setState({
        offset: Math.min(this.railWidth, this.state.offset + 1)
      });
      // pressed DOWN or LEFT arrows
    } else if (/(37|40)/.test(e.keyCode)) {
      this.setState({
        offset: Math.max(0, this.state.offset - 1)
      });
    }
  };

  render() {
    return (
      <Rail
        innerRef={rail => {
          this.rail = rail;
        }}
      >
        <Knob
          tabIndex="0"
          role="slider"
          aria-label={this.props.label}
          aria-valuemax={this.railWidth}
          aria-valuemin="0"
          aria-valuenow={this.state.offset}
          style={{
            transform: `translateX(${this.state.offset}px)`
          }}
          onMouseDown={this.handleDragStart}
          onTouchStart={this.handleDragStart}
          onTouchMove={this.handleDragMove}
          onKeyDown={this.handleKeyDown}
        />
      </Rail>
    );
  }
}

Slider.defaultProps = {
  onChange: noop
};

Slider.propTypes = {
  onChange: T.func.isRequired,
  "*": T.any
};

export default Slider;
