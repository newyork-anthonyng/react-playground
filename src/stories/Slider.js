import React from 'react';

import { storiesOf } from '@storybook/react';
import { withKnobs } from '@storybook/addon-knobs';

import Slider from '../components/Slider';
import glamorous from 'glamorous';

const Container = glamorous.div({
  margin: '50px',
});

storiesOf('Slider', module)
  .addDecorator(withKnobs)
  .add('With knobs', () => {
    return (
      <Container>
        <Slider label="Anthony" />
      </Container>
    );
  });
