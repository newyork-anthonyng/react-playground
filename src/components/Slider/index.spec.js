import React from "react";
import { mount, shallow } from "enzyme";
import noop from "../../utility/noop";
import Slider, { Knob } from "./";

const defaultProps = {
  onChange: noop
};

const map = {};
window.addEventListener = jest.fn((type, cb) => {
  map[type] = cb;
});
window.removeEventListener = jest.fn(type => {
  map[type] = undefined;
});

beforeEach(() => {
  window.addEventListener.mockClear();
  window.removeEventListener.mockClear();
});

it("should add event listeners on mount", () => {
  const wrapper = mount(<Slider {...defaultProps} />);

  // TODO: Why is there a "scroll" event listener?
  expect(window.addEventListener).toHaveBeenCalledTimes(3);
  expect(window.addEventListener.mock.calls).toMatchSnapshot();
});

it("should remove event listeners on unmount", () => {
  const wrapper = mount(<Slider {...defaultProps} />);

  wrapper.unmount();

  expect(window.removeEventListener).toHaveBeenCalledTimes(2);
  expect(window.removeEventListener.mock.calls).toMatchSnapshot();
});

it("should do nothing if mouse is moving, and not dragging", () => {
  const wrapper = mount(<Slider {...defaultProps} />);

  map.mousemove({ clientX: 100 });

  expect(wrapper.find(Knob).props().style).toMatchSnapshot();
});

it("should move the slider when dragging", () => {
  const wrapper = mount(<Slider {...defaultProps} />);
  wrapper.instance().railWidth = 300;
  const knob = wrapper.find(Knob);

  knob.simulate("mousedown", { clientX: 100 });
  map.mousemove({ clientX: 150 });

  expect(wrapper.find(Knob).props().style).toMatchSnapshot();
});

it("should move the slider on touch", () => {
  const wrapper = shallow(<Slider {...defaultProps} />);
  wrapper.instance().railWidth = 300;
  const knob = wrapper.find(Knob);

  knob.simulate("touchstart", {
    touches: [{ clientX: 100 }]
  });
  knob.simulate("touchmove", {
    touches: [{ clientX: 150 }]
  });

  expect(wrapper.find(Knob).props().style).toMatchSnapshot();
});

it("should stop moving the slider on mouse up", () => {
  const wrapper = mount(<Slider {...defaultProps} />);
  wrapper.instance().railWidth = 300;
  const knob = wrapper.find(Knob);

  knob.simulate("mousedown", { clientX: 100 });
  map.mousemove({ clientX: 150 });

  map.mouseup();
  map.mousemove({ clientX: 300 });

  expect(wrapper.find(Knob).props().style).toMatchSnapshot();
});

it("should have a minimum value of 0 when dragging slider", () => {
  const wrapper = mount(<Slider {...defaultProps} />);
  wrapper.instance().railWidth = 300;
  const knob = wrapper.find(Knob);

  knob.simulate("mousedown", { clientX: 100 });
  map.mousemove({ clientX: 0 });

  expect(wrapper.find(Knob).props().style).toMatchSnapshot();
});

it("should have a maximum value when dragging slider", () => {
  const wrapper = mount(<Slider {...defaultProps} />);
  wrapper.instance().railWidth = 300;
  const knob = wrapper.find(Knob);

  knob.simulate("mousedown", { clientX: 0 });
  map.mousemove({ clientX: 400 });

  expect(wrapper.find(Knob).props().style).toMatchSnapshot();
});

describe("Key press", () => {
  it("should move the slider when pressing UP", () => {
    const wrapper = mount(<Slider {...defaultProps} />);
    wrapper.instance().railWidth = 300;
    const knob = wrapper.find(Knob);

    knob.simulate("keydown", { keyCode: 38 });

    expect(knob.props().style).toMatchSnapshot();
  });

  it("should move the slider when pressing RIGHT", () => {
    const wrapper = mount(<Slider {...defaultProps} />);
    wrapper.instance().railWidth = 300;
    const knob = wrapper.find(Knob);

    knob.simulate("keydown", { keyCode: 39 });

    expect(knob.props().style).toMatchSnapshot();
  });

  it("should move the slider when pressing DOWN", () => {
    const wrapper = mount(<Slider {...defaultProps} />);
    wrapper.instance().railWidth = 300;
    wrapper.setState({ offset: 10 });
    const knob = wrapper.find(Knob);

    knob.simulate("keydown", { keyCode: 37 });

    expect(knob.props().style).toMatchSnapshot();
  });

  it("should move the slider when pressing LEFT", () => {
    const wrapper = mount(<Slider {...defaultProps} />);
    wrapper.instance().railWidth = 300;
    wrapper.setState({ offset: 10 });
    const knob = wrapper.find(Knob);

    knob.simulate("keydown", { keyCode: 40 });

    expect(knob.props().style).toMatchSnapshot();
  });
});
