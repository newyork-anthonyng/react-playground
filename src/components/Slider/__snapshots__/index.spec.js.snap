// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Key press should move the slider when pressing DOWN 1`] = `
Object {
  "transform": "translateX(9px)",
}
`;

exports[`Key press should move the slider when pressing LEFT 1`] = `
Object {
  "transform": "translateX(9px)",
}
`;

exports[`Key press should move the slider when pressing RIGHT 1`] = `
Object {
  "transform": "translateX(1px)",
}
`;

exports[`Key press should move the slider when pressing UP 1`] = `
Object {
  "transform": "translateX(1px)",
}
`;

exports[`should add event listeners on mount 1`] = `
Array [
  Array [
    "scroll",
    [Function],
    false,
  ],
  Array [
    "mousemove",
    [Function],
  ],
  Array [
    "mouseup",
    [Function],
  ],
]
`;

exports[`should do nothing if mouse is moving, and not dragging 1`] = `
Object {
  "transform": "translateX(0px)",
}
`;

exports[`should have a maximum value when dragging slider 1`] = `
Object {
  "transform": "translateX(300px)",
}
`;

exports[`should have a minimum value of 0 when dragging slider 1`] = `
Object {
  "transform": "translateX(0px)",
}
`;

exports[`should move the slider on touch 1`] = `
Object {
  "transform": "translateX(50px)",
}
`;

exports[`should move the slider when dragging 1`] = `
Object {
  "transform": "translateX(50px)",
}
`;

exports[`should remove event listeners on unmount 1`] = `
Array [
  Array [
    "mousemove",
    [Function],
  ],
  Array [
    "mouseup",
    [Function],
  ],
]
`;

exports[`should stop moving the slider on mouse up 1`] = `
Object {
  "transform": "translateX(50px)",
}
`;
